src/main/java/supercoder79/chunkpregen/PregenerationTask.java

```diff
@@ -70,7 +85,8 @@ public void run(Listener listener) {
         }
 
         this.listener = listener;
-        this.tryEnqueueTasks();
+
+        this.server.submit(() -> CompletableFuture.runAsync(this::tryEnqueueTasks));
```

Comment:
We're already on the server thread when `run` is called - but this will schedule onto the server thread again, and then onto the Java thread pool - is that intended? `tryEnqueueTasks` isn't safe to run from off-thread

Reply:
What I was finding is if I tried calling the pregen command in center of a large already-pregenned area, collectChunks will stall the server as it keeps checking chunks until it gets 32 (BATCH_SIZE) ungenerated chunks. If I change collectChunks to not do a continue when skipping pre-generated chunks, the queue stays empty and breaks the pregen command (seems like it thinks it ended early due to empty queue?)

So my thought was the chunk collecting can be done on the off-thread so that the server thread doesn't lock up when the game is scanning and skipping chunks for the queue. (Though I do struggle with figuring out proper ways to do off-thread queuing in MC with MC's own systems...)

Isn't tryEnqueueTasks threadsafe due to the synchronized on the queue itself? Or is the chunk scan itself the part not threadsafe?

Reply:
Oh it's the chunk ticket adding and removing that isn't threadsafe hmm. Will be tricky to deal with. Gonna have to think some more about this on how to skip existing chunks without causing the server to stall

Reply:
Sorry for all of my comments but it looks like in **enqueueChunks**, it creates a completableFuture in **whenComplete** on the **getChunk** call. In the BiFunction for that, it calls **acceptChunkResult** which then calls **tryEnqueueTasks**. 

So doesn't that mean that **tryEnqueueTasks** is already being ran on the off-threads? It's only **enqueueChunks** that needs to be put on the server thread due to the chunk tickets but the chunk scan in **tryEnqueueTasks**'s **collectChunks** can stay on the off-thread 

(haven't been able to get that off-thread collectChunks to crash in my testing. Bolded method name for myself here so I can read my own comment easier)